Fix `TSPlayer.SendTileSquare` not being centered.

0x3fcf1bbd opened the pull request:
STS has been used while expecting it to be a square whose center is the passed (x,y) coordinates, while said coordinates where used as the top left corner of the square; and this should fix it.
Hopefully I have checked enough times not to have any other oversights.

0x3fcf1bbd committed:
Fixed STS not being centered.

Inferring from the usage of it, this method was expected by its callers to be a square centered around the given coordinates (x, y), hence this being a fix.

0x3fcf1bbd committed:
Updated CHANGELOG.md

Arthri commented on TShockAPI/TSPlayer.cs:
```diff
@@ -1311,7 +1311,7 @@ public void RemoveProjectile(int index, int owner)
 		public virtual bool SendTileSquare(int x, int y, int size = 10)
 		{
-			return SendTileRect((short)x, (short)y, (byte)size, (byte)size);
+			return SendTileRect((short)(x-(size/2)), (short)(y-(size/2)), (byte)size, (byte)size);
```
I think it would be better to correct the specific use case, in this case. places like `TSPlayer.Teleport` rather than the method.

Fixing it at this method would mean a breaking change for plugins that expect it to not be centered

0x3fcf1bbd replied:
Most uses I've seen inside TShockAPI is expecting it to be centered, such as in `/grow` or `/sync`, or `TSPlayer.Teleport` as you said. 
Wouldn't it be better in the long term to have this fixed, even if it causes breaking changes to some plugins, rather than keeping a somewhat faulty method whose action can be understood in two different ways? Even then the fix for those plugins would only be to switch to STR.

bartico6 replied:
the issue is you checked usages in tshock only whereas many plugin authors may have used this before, checked how it works, and adapted to it. Such a change is potentially breaking to plugins and shouldn't be snuck into main dev like this without an API tick or whatever.

Personally I'd fix the places in tshock that use it wrong and add a second overload named SendTileSquareCentered or something that will do what your current patch suggests.

0x3fcf1bbd replied:
I guess this could work, but then should a doccomment be added to the method that states not to use it expecting a centering logic?

QuiCM replied:
Yep. Preferably we keep backwards compat but add a way to do it properly - then add an obsolete attribute to the method that is incorrect so we can remove it in future

0x3fcf1bbd replied:
Unable to think of a proper name for the other method, would an overload work? It would use a `byte` for size so to make it more obvious that this method cannot sync too large of tile areas. The current STS method having an obsolete attribute telling to switch to either the centered method or STR.
Also, should the method clamps the coordinates to avoid tile array OOBs?

hakusaro replied:
Can you overload on types with the same signature in C#? I think that's not valid in Java, but I could be wrong about which language doesn't allow that.

How about `SendTileCenteredSquare` or something? It doesn't necessarily have to be perfect. I _would_ advocate changing TShock usages to the new method though.

0x3fcf1bbd replied:
Well every name I could think of all sounded a quite awkward to me, so this is why I asked. 
I'm not sure if it is fully possible to do that, as assemblies already compiled referencing the current method should have the signature stored, and as each would differ it shouldn't break.
However, the method calls would be ambiguous due to the fact that numbers can easily be converted, so there would have to be quite a few funky casts to do so it works correctly. The obsolete attribute would still give off the warning so people would definitely know which method would be called. 
Not the best, so guess `SendTileCenteredSquare` will do.

hakusaro replied:
I think I would change my vote to putting Centered at the end or after Send, since it’s an adjective that modifies a TileSquare, actually. Other than that it’s eh.

If you can demonstrate overloading purely on types I’m happy to say that’s a good path, but I dunno what other people think.

0x3fcf1bbd replied:
Got it!
It would likely mean errors on every call unless the type of the number is explicit, so maybe not the best.

bartico6 replied:
> Well every name I could think of all sounded a quite awkward to me, so this is why I asked. I'm not sure if it is fully possible to do that, as assemblies already compiled referencing the current method should have the signature stored, and as each would differ it shouldn't break. However, the method calls would be ambiguous due to the fact that numbers can easily be converted, so there would have to be quite a few funky casts to do so it works correctly. The obsolete attribute would still give off the warning so people would definitely know which method would be called. Not the best, so guess `SendTileCenteredSquare` will do.

the problem isn't physically breaking methodcalls, the problem is that the *behaviour* of the method changed. you would now be performing a different operation under the same method, this breaks expectations.

I'd expect all overloads of the same method to act the same but take different parameters - one method taking a byte and therefore performing centered networking and not cornered networking would be a TERRIBLE developer experience to debug

bartico6 replied:
all existing calls to this method use cornered networking and I'm convinced most plugins expect that, if you're going to change it to centered, suffixing the method name + obsoleting the old one would make the most sense imo

hakusaro replied:
> the problem isn't physically breaking methodcalls, the problem is that the behaviour of the method changed. you would now be performing a different operation under the same method, this breaks expectations.

Meh, if someone has to recompile, I would rather give an old method or something anyways? E.g., keep the old function around if you want the old behavior. Ehh. Idk. It's not technically breaking ABI or API compat, but I can see how spiritually it is.

hakusaro commented:
Thank you for your work on this, by the way! I really appreciate it! ✨

0x3fcf1bbd commented:
Should the methods (STR, STS, & STSC) have the coordinates clamped so as to avoid OOBs that otherwise would be annoying? I think this would be nice so people wouldn't have to check for the coordinates to be in the world themselves. Ideally I think STR could have the clamping logic.

hakusaro commented:
> Should the methods (STR, STS, & STCS) have the coordinates clamped so as to avoid OOBs that otherwise would be annoying? I think this would be nice so people wouldn't have to check for the coordinates to be in the world themselves. Ideally I think STR could have the clamping logic.

I think that’s logical to support.

0x3fcf1bbd committed:
Reverted STS change, added STSC, fixed STS/STSC usage.

0x3fcf1bbd committed:
Updated CHANGELOG.md

0x3fcf1bbd committed:
Fixed STR width & length clamping being incorrect.

The main game should however clamp it itself too before sending, so there shouldn't be issues.

0x3fcf1bbd committed:
Merge branch 'general-devel' of https://github.com/Pryaxis/TShock into Pryaxis-general-devel

0x3fcf1bbd committed:
Merge branch 'Pryaxis-general-devel' into general-devel

Fix file conflict.

0x3fcf1bbd committed:
Removed clamping as the game does it itself.

0x3fcf1bbd commented:
Should be fine for review now.

hakusaro reviewed (approved):
![lgtm_trick](https://user-images.githubusercontent.com/532647/154391327-587586e9-79c0-45f4-8f77-2c0dadf1e279.gif)
